### PR TITLE
Feature: Add vertex color alpha channel support.

### LINKED ISSUE #16403: Include vertex color alpha?
##### Description of the problem

I'm working with a set of models that use vertex color alpha. Now that threejs uses buffer geometry almost exclusively, adding in vertex color alpha is pretty simple, and it would be nice to see these changes upstream. The changes mostly involve declaring color as a vec4 instead of vec3.

```                
'#ifdef USE_COLOR',
'   attribute vec4 color;',
'#endif',
```

```
var color_fragment = `
#ifdef USE_COLOR
    diffuseColor.rgb *= vColor.xyz;
    diffuseColor.a *= vColor.a;
#endif 
`;

var color_pars_fragment = "#ifdef USE_COLOR\n\tvarying vec4 vColor;\n#endif\n";
var color_pars_vertex = "#ifdef USE_COLOR\n\tvarying vec4 vColor;\n#endif";
var color_vertex = "#ifdef USE_COLOR\n\tvColor.xyzw = color.xyzw;\n#endif";
```

##### Example Without Vertex Color Alpha
Imgur

##### Example With Vertex Color Alpha
Imgur

##### Three.js version

- [ x ] Dev
- [ ] r104
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

### PR DESCRIPTION
**Pull request by chubei**
Related issue: Fixed #16403.

**Description**

A new shader macro `USE_VERTEX_OPACITY` is added to control if the attribute `color` is a `vec3` or `vec4`.

Corresponding flag in `Material` is `.vertexOpacities`.

Does this need a new example? @donmccurdy @mrdoob @Mugen87

### CODE REVIEW

**Comment by Mugen87** [comment]
I think I prefer the name `vertexAlpha`.

Please also update the TS declaration files and the documentation.

> Does this need a new example?

It's good if at least one example demonstrates the usage. However, I would enhance an existing buffer geometry example like [webgl_buffergeometry](https://threejs.org/examples/?q=buffergeometry#webgl_buffergeometry).

**Comment by chubei** [comment]
> I think I prefer the name `vertexAlpha`.

No problem.

> Please also update the TS declaration file and the documentation.

I wasn't aware of the TS files. Will do.

The contribution guide says `please make a new PR for the appropriate doc changes`. Is that out of date?

> It's good if at least one example demonstrates the usage. However, I would enhance an existing buffer geometry example like [webgl_buffergeometry](https://threejs.org/examples/?q=buffergeometry#webgl_buffergeometry).

I'll modify that example.

**Comment by Mugen87** [comment]
> The contribution guide says please make a new PR for the appropriate doc changes. Is that out of date?

Oh, I was not aware of that line. TBH, I think it's better if the documentation is always updated with the same PR that changes code.

**Comment by chubei** [comment]
> TBH, I think it's better if the documentation is always updated with the same PR that changes code.

I actually agree. Let's break that rule in this PR :smiling_imp:.

File: src/loaders/MaterialLoader.js
Comment (Mugen87) [inline src/loaders/MaterialLoader.js:152]: > if ( typeof json.vertexAlpha === 'number' ) {

Did you copy this code from `vertexColors`? Notice that it does not make sense to use it in context of `vertexAlpha`. The if statement comes from a time where `vertexColors` was not a boolean but an enum (with values like `THREE.VertexColors`, `THREE.FaceColors`). It was later converted to a boolean. The if statement exists only for backwards compatibility reason.

File: examples/webgl_buffergeometry.html
Comment (Mugen87) [inline examples/webgl_buffergeometry.html:138]: Would not the demo be more interesting if you apply a different alpha value per face? Otherwise I wonder why the code does not just use `Material.opacity`?

File: src/loaders/MaterialLoader.js
Comment (chubei) [reply src/loaders/MaterialLoader.js:152]: Yes I did a copy. Didn't really understand the legacy code. Will change it.

File: examples/webgl_buffergeometry.html
Comment (chubei) [reply examples/webgl_buffergeometry.html:138]: How about a random alpha value for each of the triangles? Like the rgb colors.

Comment (Mugen87) [reply examples/webgl_buffergeometry.html:138]: Sounds good!

File: examples/webgl_buffergeometry.html
Comment (Mugen87) [inline examples/webgl_buffergeometry.html:171]: You should not call `Scene.remove()` when `mesh` is `undefiend`.

I suggest you remove the usage of dat.gui and thus the option to switch between vertex alpha. Just create the data like that right away.

Comment (chubei) [reply examples/webgl_buffergeometry.html]: Sure. I also updated the screenshot.

**Review by Mugen87 — DISMISSED** [review]

**Comment by chubei** [comment]
Thanks for your review @Mugen87. Are we waiting for other collaborators’ review now?

**Comment by Mugen87** [comment]
Yep. @mrdoob will have a look at this PR and merge it when everything is fine.

**Comment by Mugen87** [comment]
@chubei Do you mind updating the PR? There is now a merge conflicts since all TS type declaration files were removed.

**Comment by chubei** [comment]
Rebased and force pushed. Sorry for the delay.

**Comment by noamgat** [comment]
Wow, this looks great! We would be really happy it it was merged.

File: src/materials/Material.js
Comment (mrdoob) [inline src/materials/Material.js:22]: @Mugen87 After #20135, do you think we still need to add `vertexAlpha`?

Comment (Mugen87) [reply src/materials/Material.js:22]: This boolean controls whether the define `USE_VERTEX_ALPHA` is added to the shader or not. Depending on this define, the color attribute is `vec3` or `vec4`.

Comment (mrdoob) [reply src/materials/Material.js:22]: But we can detect that automatically:

```js
geometry.attributes.color.itemSize === 4;
```

Comment (WestLangley) [reply src/materials/Material.js:22]: Even if an attribute is present in the geometry, we do not generally force the user to use it. Hence, the material flags.

Are tangents an exception? We should be consistent in how we handle this...

Comment (mrdoob) [reply src/materials/Material.js:22]: Well... We do not generally do it because we couldn't do it. Having these flags were a way to signal the user that materials couldn't always be reused. After #20135 we can.

Comment (chubei) [reply src/materials/Material.js:22]: I agree that we should infer as many material flags as possible from geometry attributes/object flags. I’ll read #20135 related code more carefully later and see if ‘vertexAlpha’ can be inferred automatically.

We probably should start removing other flags as well? Of course not in this PR.

Comment (mrdoob) [reply src/materials/Material.js:22]: > I agree that we should infer as many material flags as possible from geometry attributes/object flags. I’ll read #20135 related code more carefully later and see if ‘vertexAlpha’ can be inferred automatically.

Sound good 👍

I think this check should work:

```js
vertexAlpha: geometry.attributes.color && geometry.attributes.color.itemSize === 4,
```

> We probably should start removing other flags as well? Of course not in this PR.

Yep! I think we should be able to remove `vertexColors`, `vertexTangents` and `skinning` 💆‍♂️

Comment (Mugen87) [reply src/materials/Material.js:22]: @mrdoob Although #20135 is merged, it does not mean materials properties like `vertexAlpha` are not necessary anymore. If we don't have these flags, we need code like in `setProgram()`:

https://github.com/mrdoob/three.js/blob/6cfe91706d1b5434f250d7be76cad29287926f58/src/renderers/WebGLRenderer.js#L1490-L1498

These checks are important since we only want to enter `getProgram()` when required.

Comment (WestLangley) [reply src/materials/Material.js:22]: @WestLangley wrote:

>Even if an attribute is present in the geometry, we do not generally force the user to use it.

I do not think we should _infer_ material properties from the _geometry_.

Comment (chubei) [reply src/materials/Material.js:22]: > I do not think we should infer material properties from the geometry.

@WestLangley Do you mind to explain the motivation?

For my motivation for inferring the flags, imagine the following case. A user may set the `color` attribute to be a 4-component Uint8Array, containing the R,G,B,A,R,G,B,A... values, but forget to set the `vertexAlpha` flag. WebGL will then silently discard the `A` values and the user may have to dig into `MeshStandardMaterial` and further `Material` documentation to see why. Inferring the `vertexAlpha` flag can avoid this case.

Comment (WestLangley) [reply src/materials/Material.js:22]: @chubei We can't prevent user errors.

The following code would not be possible if material properties were inferred from the geometry:
https://github.com/mrdoob/three.js/blob/6deef54a1ece0fef56765885b19044db27469f04/editor/js/Viewport.js#L46-L48

Comment (chubei) [reply src/materials/Material.js]: > We can't prevent user errors.

I agree that we can't prevent all of them, but we can prevent some of them, making the API easier to use.

`grid1.material.vertexColors = false; ` can be changed to `grid1.geometry.deleteAttribute('color');` for equivalent effect. Of course compatibility is a problem. I guess we have different opinions on whether it's worth it.

Anyway, this feature actually doesn't need either manual or automatic `vertexAlpha` flag. I just realized if you pass 3-component values to WebGL, it'll fill the last component with the default value for that attribute, which is default to 1.0. We can always use `vec4` in the shader.

Comment (Mugen87) [reply src/materials/Material.js]: Sorry, but I do not vote to implement vertex color alpha like that since it's a breaking change. Custom shader code might break with this message:

> cannot convert from 'in highp 4-component vector of float' to 'out highp 3-component vector of float'

Comment (chubei) [reply src/materials/Material.js]: Didn't realize it's a breaking change. Thanks for pointing that out @Mugen87. I'll reset to last reviewed state and continue investigating automatic inferring.

Comment (Mugen87) [reply src/materials/Material.js]: @Do you mind resetting the PR to 07f1cb16a2e438a5b771415a442fb01a081d5b2b?

**Review by Mugen87 — CHANGES_REQUESTED** [review]
Using `vec4` by default is a breaking change and can't be accepted.

File: src/materials/Material.js
Comment (chubei) [reply src/materials/Material.js:22]: > @chubei Do you mind resetting the PR to [07f1cb1](https://github.com/mrdoob/three.js/commit/07f1cb16a2e438a5b771415a442fb01a081d5b2b)?

Of course. Reset.

**Review by Mugen87 — APPROVED** [review]

**Comment by chubei** [comment]
I tried to implement automatic inferring `vertexAlpha` and force pushed the code. Because most of the original modification is not used any longer, I rebased and made a new commit. Please let me know if you want a merge-then-forward style commit.

**Comment by Mugen87** [comment]
I think we should stick to `vertexAlpha` and let the user decide when to use vertex alpha or not.

**Review by Mugen87 — CHANGES_REQUESTED** [review]
Not approving an inferring approach.

File: src/renderers/WebGLRenderer.js
Comment (Mugen87) [inline src/renderers/WebGLRenderer.js:1517]: This is exactly the problematic line I've mentioned before which affects now all calls of `setProgram()` when a material is not changed.

**Comment by chubei** [comment]
@mrdoob What’s your opinion on this? I prefer the inferring approach, but am also fine with merging the previous version.

File: src/renderers/WebGLRenderer.js
Comment (chubei) [reply src/renderers/WebGLRenderer.js:1517]: How about adding a ‘layoutChanged’ flag in ‘geometry’, which gets set every time a new attribute is added or an attribute’s type/size is changed? If the user changed the attribute layout during rendering, he probably wanted the material being updated.

Comment (chubei) [reply src/renderers/WebGLRenderer.js:1517]: How about adding a ‘layoutChanged’ flag in ‘geometry’, which gets set every time a new attribute is added or an attribute’s type/size is changed? If the user changed the attribute layout during rendering, he probably wanted the material being updated.

Comment (Mugen87) [reply src/renderers/WebGLRenderer.js:1517]: It's not only the geometry that would requires these checks. You also have to check all kinds of 3D objects at this place. This can all be avoided by using material properties.

Comment (Mugen87) [reply src/renderers/WebGLRenderer.js:1517]: Besides, your approach does still not solve https://github.com/mrdoob/three.js/pull/20975#discussion_r589639751.

File: src/materials/Material.js
Comment (Mugen87) [reply src/materials/Material.js]: > grid1.material.vertexColors = false; can be changed to grid1.geometry.deleteAttribute('color'); for equivalent effect.

This becomes problematic if you want to reuse the same geometry with different materials. According to your  approach, you would have to clone the geometry multiple times and configure it differently. This sounds quite way less efficient than just using different materials.

File: src/renderers/WebGLRenderer.js
Comment (chubei) [reply src/renderers/WebGLRenderer.js:1517]: > It's not only the geometry that would requires these checks. You also have to check all kinds of 3D object properties. This can all be avoided by using material properties.

Had a quick look at the documentation of `Object3D` and `Objects` section. Looks like the properties affecting shader are only `geometry` and `receiveShadow`? Sorry if missed anything.

File: src/materials/Material.js
Comment (chubei) [reply src/materials/Material.js]: > This becomes problematic if you want to render the same geometry with different effects. According to your approach, you would have to clone the geometry multiple times and configure it differently. This sounds less efficient than just using the same geometry but different materials.

Creating a new `geometry` without cloning underlying `TypedArray` should still be quite cheap? Not saying it’s better than material flags in terms of efficiency, just discussing possibilities.

Comment (Mugen87) [reply src/materials/Material.js]: Even when performing a shallow clone the workflow still sounds strange to me.

**Comment by mrdoob** [comment]
Hmm...

On the API ergonomics side, inferring is a big win.
Users no longer need to know about `skinning`, `vertexTangents`, `vertexColor`, `vertexAlpha`, etc

We can also remove 3 checkboxes from the editor (which is usually a good signal API wise too):

Screen Shot 2021-03-15 at 5 51 25 PM

I agree with @chubei that instead of `grid1.material.vertexColors = false;` the user should be doing `grid1.geometry.deleteAttribute('color');` (and I should have done that 😅).

I also understand that this would force people have to duplicate geometries and that could result in increased memory, but I'm not sure how much of a concern it should be for most of the cases.

How about we try the inferred approach in this case so we can see what the real consequences are?
We can add the property to the material later if we found this creates issues.

**Comment by Mugen87** [comment]
> How about we try the inferred approach in this case so we can see what the real consequences are?

Okay, that sounds good to me.

File: src/renderers/WebGLRenderer.js
Comment (Mugen87) [reply src/renderers/WebGLRenderer.js:1517]: That list is not complete, sorry.

**Review by Mugen87 — DISMISSED** [review]

File: src/core/BufferGeometry.js
Comment (Mugen87) [inline src/core/BufferGeometry.js:882]: I don't think we want to expose this method.

Comment (chubei) [reply src/core/BufferGeometry.js:882]: Yes, this also looks strange to me. Any suggestions where I should put this method? Don't feel like to repeat the code in `WebGLRenderer.js` and `WebGLPrograms.js`.

Comment (mrdoob) [reply src/core/BufferGeometry.js:882]: Agreed! I was going to clean that up after merging it tomorrow.

**Comment by mrdoob** [comment]
@Mugen87

I've spent some idle thinking on this issue over the last few days and I came up with another reason to remove `skinning`.

There are quite a few bug reports of `scene.overrideMaterial` not working with scenes that have skinned and not skinned meshes. And that's because we can't reuse the same material in both. If a user wanted to render the whole scene in wireframe for example, it's not possible.

Similarly, `WebGLShadowMap` has [this logic](https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLShadowMap.js#L231-L277) that is a workaround for the same problem.

My hunch is that the extra VAOs are worth it so users can finally reuse materials without bumping into these issues and being forced to understand the internals.

File: src/core/BufferGeometry.js
Comment (chubei) [reply src/core/BufferGeometry.js]: > I would indeed inline it for now.

Sure thing. Updated.

**Comment by mrdoob** [comment]
Hmm, I'm going to merge this and do another PR some changes to it for your review.

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**
